pyproject.toml

```diff
@@ -45,6 +45,7 @@ dependencies = [
     "setuptools-scm>=7.1.0",
     "torch>=1.13.1,<3",
     "typing_extensions>=4.7.0",
+    "pyarrow>=14.0.0",
```

Comment:
Can you comment on why this is necessary?

Reply:
Core tests didn't pass due to missing dependency.

Reply:
@AdrianSosic does this relate to the one thing you found at some time? I remember you saying that there was some dep installed/pinned only in fulltest and not core

Reply:
I think what you are referring to is a different issue that was related to mixing our dependency groups. See commit 227182d7ee3e2e28ed2254292e102f02bd7fba64 (message: `Remove "lint" from "test" dependency group`). But that had nothing to do with the current one. Can you confirm?

I think the `pyarrow` here is probably required by some special `polars` feature, perhaps the `streaming`? Does someone of you know, perhaps? But anyway, I think it should thus be installed as a secondary dependency through `polars`, i.e. by replacing `polars` with `polars[pyarrow]` in our toml file

Reply:
It was required for `join`, but I don't know exactly where.

Reply:
Ok, so can we then resolve this? Seems like this is then something we need to include ad have no choice, right?

Reply:
No, for two reasons:
* We should understand why/if pyarrow is really needed. For that, someone needs to conduct a test in an environment where it is **not** installed and determine the operation where it is needed.
* If it turns out it is really needed for polars, then we should change it into a polars secondary dependency, instead of a primary one

Reply:
Polars `.to_pandas` seems to require `pyarrow`, see [here](https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.to_pandas.html)
<img width="546" alt="image" src="https://github.com/user-attachments/assets/0b94d303-0487-4b8f-9500-769a912c8e97">
seems pretty unavoidable to me

Reply:
ok, seems to be avoidable by changing the corresponding searchspace creation step from
```python
pl_df = lazy_df.collect(streaming=True)
df = pl_df.to_pandas()
```
to 
```python
df_dict = lazy_df.collect(streaming=True).to_dict()
df = pd.DataFrame.from_dict(df_dict)
```
or
```python
df_dicts = lazy_df.collect(streaming=True).to_dicts()
df = pd.DataFrame.from_records(df_dicts)
```
(Haven't tested any impact on speed or memory)
Would that be fine @AdrianSosic ? If so we can avoid pyarrow dep

Reply:
Hey @Scienfitz, I would not cut down here on code on performance. For me, it's absolutely OK to have pyarrow as a transitive dependency – it was only important to me to be certain that we really need it. So keep the code as is, but please change it into a proper transitive dep by moving it to the polars entry 👍🏼

Reply:
im confused, its pretty clear that we dont need it, so why do you want to use it now? We dotn know anything about performance

Reply:
The point is here that you go via dict conversion, which is definitely suboptimal. The pyarrow approach does it directly, even avoiding copies (see documentation). So it's clear that the dict approach does not require pyarrow since you basically deserialize your data, but that's also exactly where the downside lies

Reply:
In a sense, pyarrow gives you a direct polars <-> pandas converter, while your solution makes the de-/serialization detour

Reply:
so to be clear you do *not* want to avoid the additional dependency (even if its introduced via `polars[pyarrow]`)? Because thats the vibe i got from previous comments

Reply:
What? No, I *do want* to have it as `polars[pyarrow]`. Btw, another reason agains going via dicts, you can lose type information (also mentioned in the docs). 

And here you have the speed black on white, pretty convincing I think :D 
Speed: ~1s vs ~125s
Benchmark script:
```python
from time import perf_counter
import pandas as pd
import polars as pl
import numpy as np

df = pl.DataFrame(np.random.random((10000, 1000)))
t = perf_counter()
for _ in range(100):
    df.to_pandas()
    # pd.DataFrame.from_records(df.to_dicts())
print(perf_counter() - t)
```

Reply:
ok that is a pretty obv benefit, the confusions stems from this comment 
![image](https://github.com/user-attachments/assets/a896c51e-7e3d-4194-b2bf-e8ace33151cf)
 which doesnt mention speed at all and sounds like `whether its essential`. Its clearly not essential (without the previously not mentioned speed requirement). 

In any case, I updated this PR and will merge it now, if theres any further discusison needed it can be done in #318 and subsequent

Reply:
Yes, when I wrote that statement, I was not thinking of the possibility of hacky workarounds 😄 Great, thx for the fix